cluster/zookeeper/locks.go

```diff
@@ -0,0 +1,74 @@
+package zookeeper
+
+import (
+	"fmt"
+	"sort"
+)
+
+// LockEntries is a container of locks.
+type LockEntries struct {
+	// Map of lock ID integer to the full znode path.
+	m map[int]string
+	// List of IDs ascending.
+	l []int
+}
+
+// locks returns a LockEntries of all current locks.
+func (z *ZooKeeperLock) locks() (LockEntries, error) {
+	var locks = LockEntries{
+		m: map[int]string{},
+		l: []int{},
+	}
+
+	// Get all nodes in the lock path.
+	nodes, _, e := z.c.Children(z.Path)
+	// Get the int IDs for all locks.
+	for _, n := range nodes {
+		id, err := idFromZnode(n)
+		// Ignore junk entries.
+		if err == ErrInvalidSeqNode {
+			continue
+		}
+		// Append the znode to the map.
+		locks.m[id] = fmt.Sprintf("%s/%s", z.Path, n)
+		// Append the ID to the list.
+		locks.l = append(locks.l, id)
+	}
+
+	sort.Ints(locks.l)
+
+	return locks, e
+}
+
+// IDs returns all held lock IDs ascending.
+func (le LockEntries) IDs() []int {
+	return le.l
+}
+
+// First returns the ID with the lowest value.
+func (le LockEntries) First() (int, error) {
+	if len(le.IDs()) == 0 {
+		return 0, fmt.Errorf("no active locks")
+	}
+
+	return le.IDs()[0], nil
+}
+
+// LockPath takes a lock ID and returns the znode path.
+func (le LockEntries) LockPath(id int) (string, error) {
+	if path, exists := le.m[id]; exists {
+		return path, nil
+	}
+	return "", fmt.Errorf("lock ID doesn't exist")
+}
+
+// LockAhead returns the lock ahead of the ID provided.
+func (le LockEntries) LockAhead(id int) (int, error) {
```

Comment:
I guess it's safe to say that this will never receive `id=1` as parameter given that zookeeper will never provide the same sequential ID to two simultaneous `CreateProtectedEphemeralSequential` request and any znode with ID=1 will get the lock for sure. 

Asking as this is the only case this function will panic with index out of range.

Reply:
Yeah, ZooKeeper provides linearizable writes and `sequential` znodes function as atomic counters: clients will always get a unique id and in order according to request arrival at the ZK side